numpy/core/include/numpy/ndarraytypes.h

```diff
@@ -701,6 +701,7 @@ typedef struct tagPyArrayObject_fields {
     int flags;
```

Comment:
Could we use flags to version this struct and indicate whether `_buffer_info` exists or not? If it doesn't exist we should raise an error when trying to use it.

Reply:
I am not sure I am following.  We could add a version, but I guess we already have the NumPy version to find out how large the struct will be (or even the type object).  If we would version it, C-side subclassers (if we can even figure out someone is that), would need to opt-in with a flag and we otherwise keep using the global? (and maybe try to give a deprecation warning)

Reply:
I was thinking more along the lines of "Detect that we did not set up this `PyArrayObject_fields`, refuse to do any buffer processing". That would force anyone mimicking our struct to also add the field and update the flag, otherwise they would get errors. So then there are two parts: 
- set something inside the struct itself to show it has this extra field
- detect this thing and raise an error rather than write past the end of the struct

It needs to be inside the struct since the struct can be fed in by C-API consumers.

Reply:
I would be happy to get rid of the global registry and force the few people who reproduce the `PyArrayObject_fields` struct to update or fail.

Reply:
Good point, I have to think whether its feasible/easy, since python subclasses must not notice any of such things.

Reply:
@mattip, I guess we could initialize the field with a singleton/garbage (instead of NULL), in the hope that it makes clashes/crashes more likely.  The only usage that would really find, is if the subclasser also initializes the field to NULL and does not end up using it consistently.  (Since any non-NULL value should lead to a crash during dealloc.)

I am not sure there is a good way to find C-side subclassing, aside from checking whether the types `basicsize` changed, which would be the case unless the subclasser added a single pointer at the end.  So that is something we can do: raise an error if the types size differs, but there isn't some additional flag set.  But it seems a bit convoluted? We would have to set that flag on each instance probably, or require reimplementation of the buffer export, which seems strange?

Reply:
This is the scenario that worries me:
- Someone uses the older C-API to create an ndarray, using something like `PyArray_NewFromDescr`. Now they have a `PyArrayObject` that is smaller than the one after this PR.
- They somehow pass it to a newer NumPy. Now we get a `PyObject*`, and 
- NumPy 1.20 casts it to a `PyArrayObject`. Whoops, we will read this field off the end of the memory they gave us.

This is similar to the problems we were having with Cython compiled code like cython/cython#2221 that led to cython/cython#2627.

`sizeof()` or junk memory will not help here. I think the only way to detect this is to have some kind of versioning in the struct itself, which is why I suggested we add a value to the flag, I think something like `#define NPY_ARRAY_HAS_BUFFER_INFO 0x4000` might work. Then we can set the flag whenever we allocate and initialize a new `PyArrayObject_fields`, and check that the flag has that value set whenever we wish to manipulate the `_buffer_info`.

Reply:
`basicsize` might work as well.

Reply:
So the example we want to guard against is mainly this type of thing?:
```
cimport numpy as npc

cdef class sub(npc.ndarray):
    pass
```
without any extra fields the struct will be too small (I thought cpython would check this, but it does not), if we add more fields they will be incompatible.  The case where the struct is too small is simple though: We can just raise an error in `tp_alloc`.  But if you add a single new field there, we will not know and need the flag (or hope for the best with guards).
I am actually not sure where the flag should be set. `tp_alloc` is plausible, but not sure its available in cython, I guess in cython you may have to set it in `__array_finalize__` (but only if numpy is new enough).

For heaptypes, we would have to walk the mro (to see if there is a static subclass involved) – they always have at least one additional field `tp_dict`, and may have more for `__slots__`. Or simply assume that there is no static subclass involved.

I am not sure about `PyArray_NewFromDescr`, you can't create an incompatible `ndarray` in any reasonable way I think. (It would have to be a stack allocated array, which is possible, but pretty dirty – at least outside of NumPy).

Reply:
Let's discuss the implications at the next triage meeting.

Reply:
OK, thinking about it a bit more. I don't have a big new idea.  We noted flags in the meeting, there is the option to:

1. Add a flag to indicate that a buffer was exported, the only use would be a sanity check on `dealloc` (so that we do not crash hard).
2. Somehow allow downstream to indicate that they are aware of a growing struct. But I don't see how downstream can add a flag here, since it needs to be flagged on the type object.

There is no perfect solution, since point 1 can easily crash hard if an array exports a buffer and then the field gets modified.  That is super rare, but I thought an alternative could be to tag the pointer to give us a chance to convert the hard crash into a useful message.

I was going to polish this up a bit, but it would be good to finish gh-16936 and gh-17295 since it modifies the same code.

Reply:
Isn't [the `PyArray_GetNDArrayCVersion` mechnism](https://numpy.org/doc/stable/reference/c-api/array.html?highlight=pyarray_getndarraycversion#c.PyArray_GetNDArrayCVersion) supposed to adequately protect us from ABI changes?

Reply:
Ooops, yeah. That forces everyone to recompile even though almost nobody will have to, but is maybe best. But, it seems the current check won't allow to easily compile against both versions of NumPy, which is probably still problematic (unless we assume that this happens so rarely that packaging isn't probably an issue for these projects)?

Reply:
Maybe try building this PR and running it against an version of SciPy compiled for 1.19 or older

Reply:
I think PyArray_GetNDArrayCVersion works properly when mixing NumPy versions in, say, SciPy and Numpy. Otherwise we would have heard loud screams when we modified PyUFuncObject in v1.16 and changed the ABI version number. So I am 99% sure this discussion is a non-issue. I guess one way to find out is to merge this now, and wait for the weekly builds to percolate through the system.

Reply:
... which means fixing the merge conflict and getting tests to pass.

Reply:
I hope it is, but check here:

https://github.com/numpy/numpy/blame/39c915c1236d61debb4b3e10dfadf910da6c9b30/numpy/core/setup_common.py#L24

We have the API version and the ABI version. The API version we can increase easily, since it will only affect anything if you compile against a new NumPy version and then run it on an old one (which we can assume nobody does). But the ABI version has not been touched in 10 years.  Now I don't know how well that worked out 10 years ago, but it seems like a time before widespread comfortable pip?

Reply:
OK, then I guess we need to do some testing to make sure.

Reply:
I played with this for a while with `pip install scipy` and a local `numpy` build (after changing `C_ABI_VERSION` in `numpy/core/setup_common.py`. I also modified `_import_array` in `b/numpy/core/code_generators/generate_numpy_api.py` to print when it is called. If I import `scipy`, I see it calls `_impport_array`, and the `C_ABI_VERSION` is the new one. Which makes sense, the function is exposed via `_multiarrayumath.so`, which comes from NumPy, but I just wanted to confirm.

Reply:
I don't understand.  I just tried as well. First, I had to clean up to get the new C-ABI version, I think.  Second `import_array` is defined as part of `__multiarray_api.h` and should be included as a header and not linked? If it is linked, that would defeat the purpose.  So I assume your pip install compiled scipy?

Importing the system ScPy after incrementing the C-ABI version, I get this (version was 100009 or so before):
```
In [2]: f"{np.core._multiarray_umath._get_ndarray_c_version():x}"                                                       
Out[2]: '200000a'

In [3]: from scipy import linalg                                                                                        
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
RuntimeError: module compiled against ABI version 0x1000009 but this version of numpy is 0x200000a
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
ImportError: numpy.core.multiarray failed to import
```

Reply:
OK, I didn't try importing scipy.linalg (which I guess is compiled by cython). Thanks for checking more deeply. So we cannot change C_ABI_VERSION without recompiling the world.

Where does that leave us with this PR (and the one to configure memory allocations)?

Reply:
I guess we either:

1. Wait until some day where we actually do a big bang NumPy 2.0 (I am fan of a 2.0, but not of recompile the world 2.0)
2. Define the size of these structs as "undefined"? Placing it into the downstream responsibility of checking for binary compatibility in this case (this is what cython normally does, although with a warning? I have yet to test it here though).
3. We think about some extremely slow route, with compile time warnings...?

My opinion about option 2 has not changed (with two caveats):

* I would like to give a reasonable error/warning at least in the majority of (hypothetical) cases. This is to *decrease the severity* if the issue appears.
* No big "user" of this shows up and shouts loudly. Which is, unfortunately, plausible.

Of course, if you go by strict policies, you might just dismiss my opinion. But IMO, we should be using quality assurance type triaging, which means map out how likely it is (how many users are affected), how sever it is:

* Affects: Exceedingly few users (we are not aware of any, none of the large open source pages seem affected)
* Severity: Crash bug, which is easy to fix by recompiling once the issue is found.  That is fairly sever, but super high, since it is seems much impossible that you get wrong results without a loud crash as well.

So because it should be reasonable to fix, and be a loud crash that affects very few users. My personal opinion is that it might well be much lower impact than some "bug fixes".

Of course, you have to weigh that with the benefit. In this case probably a few percent of speedup for many scripts relying on small to mid-sized arrays and simpler code.  And later your change, which might make some big-data  use cases much faster.

Reply:
On the up-side here, I was trying around with cython code such as:
```
cdef class myarr(np.ndarray):
    cdef int myint  # or just pass
```
And even that crashes hard. I guess because cython supports cyclic GC, but we do not.  So that rules out one set of potential users who run into this unwittingly...

I wonder if could declare `np.ndarray` as final in the `__init__.pyd`. Or in some other way telling cython that `cdef` style subclassing is *not* valid (or at least not simple).